### PR TITLE
[WIP] Better Future

### PR DESCRIPTION
**Pull request by zielmicha**
(For issue #5731)

* [x] Split asyncfutures into its own module
    This slightly changes behaviour of callSoon - before loop is initialized, callSoon will call the function immediately.
* [x] Move `FutureStream` to separate module
* [ ] make Future fat-pointer
* [x] Callbacks in linked list

### CODE REVIEW

File: lib/pure/asyncfutures.nim [hunk 1]
```diff
@@ -30,7 +33,14 @@ type
 when not defined(release):
   var currentID = 0
 
-proc callSoon*(cbproc: proc ()) {.gcsafe.}
+var callSoonProc* {.threadvar.}: (proc(cbproc: proc ()) {.gcsafe.})
+
+proc callSoon*(cbproc: proc ()) =
+  if callSoonProc == nil:
+    # Loop not initialized yet. Call the function directly to allow setup code to use futures.
```
Comment (yglukhov) [inline lib/pure/asyncfutures.nim]: Do I understand correctly that this has basically the same issue that #5741 has. In a background thread globalDispatcher might never be inited? Maybe it's better to `getGlobalDispatcher` here?

Comment (zielmicha) [inline lib/pure/asyncfutures.nim]: Yes, it has. But I think it's unlikely to hit this bug before initializing the dispatcher. If the dispatcher is not initialized in some other place, `callSoon` will anyway never be called.

I would like to avoid making `asyncfutures` depend on `asyncdispatch`. My "ultimate goal" is making event loops pluggable, in a similar manner to Pythons asyncio.

Comment (zielmicha) [inline lib/pure/asyncfutures.nim]: E.g. `waitFor` initializes the global dispatcher.

Comment (zielmicha) [inline lib/pure/asyncfutures.nim]: (Note that GlobalDispatcher is a misleading name, it actually should be called ThreadDefaultDispatcher)

Comment (yglukhov) [inline lib/pure/asyncfutures.nim]: Ok, LGTM.

File: lib/pure/asyncfutures.nim [hunk 2]
```diff
@@ -30,7 +34,14 @@ type
 when not defined(release):
   var currentID = 0
 
-proc callSoon*(cbproc: proc ()) {.gcsafe.}
+var callSoonProc* {.threadvar.}: (proc(cbproc: proc ()) {.gcsafe.})
```
Comment (dom96) [inline lib/pure/asyncfutures.nim]: I think it would make sense to follow the same convention as the dispatcher's:

* A ``setCallSoonProc`` procedure, and
* a ``getCallSoonProc`` procedure

File: lib/pure/asyncfutures.nim (hunk 1 above)
Comment (dom96) [inline lib/pure/asyncfutures.nim]: I feel like interfaces would help us out a lot for this pluggable goal. Maybe we should wait until they stabilise?

File: lib/pure/asyncfutures.nim [hunk 3]
```diff
@@ -1,8 +1,18 @@
+include "system/inclrtl"
+
+import os, tables, strutils, times, heapqueue, options, deques
 
 # TODO: This shouldn't need to be included, but should ideally be exported.
 type
+  CallbackFunc = proc () {.closure, gcsafe.}
+
+  CallbackList = object
```
Comment (zielmicha) [inline lib/pure/asyncfutures.nim]: I don't think it's possible - the first item needs to be inlined into `FutureVar` (to save alloc in common case).

File: lib/pure/asyncfutures.nim (hunk 2 above)
Comment (zielmicha) [inline lib/pure/asyncfutures.nim]: Ok.

Interfaces won't help here - asyncfutures only needs this single callSoon proc and I think it shouldn't depend on whole AsyncLoop interface. (anyway, I think there are better options than using interface for AsyncLoop, but that's another topic)

**Comment by yglukhov** [comment]
@zielmicha what's the status of this? I would really appreciate separation of futures.

**Comment by zielmicha** [comment]
This mostly, works apart for the few problems mentioned by @dom96. "make Future fat-pointer" turned out to be a bit more problematic.

I'll try to finalize the code soon.

**Comment by zielmicha** [comment]
The new commit addresses issues raised by @dom96.

**Comment by yglukhov** [comment]
So, good to go? @zielmicha, @dom96

**Comment by zielmicha** [comment]
For me it's good to go.

File: lib/pure/asyncfutures.nim [hunk 4]
```diff
@@ -30,7 +32,27 @@ type
 when not defined(release):
   var currentID = 0
 
-proc callSoon*(cbproc: proc ()) {.gcsafe.}
+var callSoonProc {.threadvar.}: (proc(cbproc: proc ()) {.gcsafe.})
```
Comment (dom96) [inline lib/pure/asyncfutures.nim]: Also, are the parenthesis around the proc type necessary?

File: lib/pure/asyncfutures.nim [hunk 5]
```diff
@@ -98,33 +104,57 @@ proc checkFinished[T](future: Future[T]) =
       err.cause = future
       raise err
 
+proc call(callbacks: var CallbackList) =
+  var current = callbacks
+
+  while true:
+    if current.function != nil:
```
Comment (dom96) [inline lib/pure/asyncfutures.nim]: Again, ``isNil`` should be used in this proc.

File: lib/pure/asyncfutures.nim [hunk 6]
```diff
@@ -98,33 +104,57 @@ proc checkFinished[T](future: Future[T]) =
...
 
 proc complete*[T](future: FutureVar[T]) =
   ## Completes a ``FutureVar``.
   template fut: untyped = Future[T](future)
   checkFinished(fut)
   assert(fut.error == nil)
   fut.finished = true
-  if fut.cb != nil:
-    fut.cb()
+  fut.callbacks.call
```
Comment (dom96) [inline lib/pure/asyncfutures.nim]: Missing `()`. 

I bet this works anyway, but let's be consistent with the style.

File: lib/pure/asyncfutures.nim [hunk 7]
```diff
@@ -152,26 +176,36 @@ proc fail*[T](future: Future[T], error: ref Exception) =
   future.error = error
   future.errorStackTrace =
     if getStackTrace(error) == "": getStackTrace() else: getStackTrace(error)
-  if future.cb != nil:
-    future.cb()
+  future.callbacks.call
```
Comment (dom96) [inline lib/pure/asyncfutures.nim]: Same here with `()`.

File: lib/pure/asyncfutures.nim [hunk 8]
```diff
@@ -152,26 +165,36 @@ proc fail*[T](future: Future[T], error: ref Exception) =
...
+
+proc clearCallbacks(future: FutureBase) =
+  future.callbacks.function = nil
+  future.callbacks.next = nil
+
+proc addCallback*(future: FutureBase, cb: proc() {.closure,gcsafe.}) =
```
Comment (dom96) [inline lib/pure/asyncfutures.nim]: I guess that's a "yes"?

File: lib/pure/asyncfutures.nim [hunk 9]
```diff
@@ -152,26 +176,36 @@ proc fail*[T](future: Future[T], error: ref Exception) =
...
   else:
-    # This is to prevent exceptions from being silently ignored when a future
-    # is discarded.
-    # TODO: This may turn out to be a bad idea.
-    # Turns out this is a bad idea.
-    #raise error
-    discard
+    future.callbacks.add cb
+
+proc addCallback*[T](future: Future[T], cb: proc(future: Future[T]) {.closure,gcsafe.}) =
```
Comment (dom96) [inline lib/pure/asyncfutures.nim]: Spacing and word wrap (80 chars).

```nim
proc addCallback*[T](future: Future[T],
                     cb: proc (future: Future[T]) {.closure, gcsafe.}) =
```

File: lib/pure/asyncfutures.nim [hunk 10]
```diff
@@ -152,26 +176,36 @@ proc fail*[T](future: Future[T], error: ref Exception) =
...
+  ## Adds the callbacks proc to be called when the future completes.
+  ##
+  ## If future has already completed then ``cb`` will be called immediately.
+  future.addCallback proc() = cb(future)
```
Comment (dom96) [inline lib/pure/asyncfutures.nim]:
```nim
future.addCallback(
  proc () =
    cb(future)
)
```

File: tests/async/tasyncrecursion.nim [hunk 11]
```diff
@@ -17,5 +17,6 @@ proc asyncRecursionTest*(): Future[int] {.async.} =
     inc(i)
 
 when isMainModule:
+  setGlobalDispatcher(newDispatcher())
```
Comment (dom96) [inline tests/async/tasyncrecursion.nim]: Why is this necessary? The main thread should always have an implicit dispatcher.

File: lib/pure/asyncstreams.nim [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+import deques
+
+type
+  FutureStream*[T] = ref object   ## Special future that acts as
+                                  ## a queue. Its API is still
+                                  ## experimental and so is
+                                  ## subject to change.
+    queue: Deque[T]
+    finished: bool
+    cb: proc () {.closure, gcsafe.}
```
Comment (dom96) [inline lib/pure/asyncstreams.nim]: It's fine if you do not wish to implement this, but should future streams support this?

File: lib/pure/asyncdispatch.nim [hunk 13]
```diff
@@ -9,11 +9,12 @@
...
 include "system/inclrtl"
 
-import os, tables, strutils, times, heapqueue, options
-
+import os, tables, strutils, times, heapqueue, options, asyncstreams
+import asyncfutures except callSoon
 import nativesockets, net, deques
 
 export Port, SocketFlag
+export asyncfutures, asyncstreams
```
Comment (zielmicha) [inline lib/pure/asyncdispatch.nim]: Yes, it is exported (callSoon is exported from asyncfutures, so `export asyncfutures` re-exports it).

**Comment by zielmicha** [comment]
Future streams also should support multiple callbacks. I plan to implement them as part of next changes, but it make take some time.

**Comment by zielmicha** [comment]
Formatting fixed. 

Is there any utility that checks formatting for stdlib?

File: tests/async/tasyncrecursion.nim (hunk 11 above)
Comment (yglukhov) [inline tests/async/tasyncrecursion.nim]: @zielmicha, can you revert it please? The test runs just fine, creating the dispatcher lazily.

Comment (zielmicha) [inline tests/async/tasyncrecursion.nim]: Strange, I would swear it didn't work before.

**Comment by yglukhov** [comment]
Hrm... ok, my bad. The test fails on travis with stackoverflow, but works locally on my machine. I think it should be fixed, but not by changing the test. As far as I understand the "stupid" version of `callSoon` manages to initialize earlier than that of dispatcher.

**Comment by zielmicha** [comment]
If you know any good way of fixing that, let me know.

I don't think this issue may come up in real world.

**Comment by yglukhov** [comment]
@zielmicha, ok I'm sorry again, I have misunderstood the reason for this. So I agree that your initial change to the test is ok. As a slight improvement, you can add a thread local counter to the "stupid" version of `callSoon` to count the recursion, and if this counter exceeds some threshold, say 1000, it would raise an exception with a friendlier error message. @dom96, what do you think?

**Comment by dom96** [comment]
> As far as I understand the "stupid" version of callSoon manages to initialize earlier than that of dispatcher.

How can that be? What is the actual issue here? I feel like this could be hiding a bad bug.

**Comment by yglukhov** [comment]
@dom96, that my assumption is not totally accurate. The new logic of the `callSoon` is try to call `callSoonProc`, and if it is not set, fall back to recursive behavior (like it was some while ago). As soon as `getCurrentDispatcher` kicks in, it overrides the `callSoon` behavior to avoid recursion. This particular test in question tested exactly the case with deep recursion, by entering recursion on the very early stage, not letting the dispatcher to initialize. Such behavior is synthesized specifically to test the dispatcher logic, and is unlikely to be reproduced in real life, when dispatcher will be initialized with the first `poll` call.

**Comment by dom96** [comment]
IMO the initialization should be implicit. So when ``callSoon`` is called it should initialize the dispatcher.

**Comment by yglukhov** [comment]
True, but that is not possible without a forward declared dispatcher interface, which I think we're yet to come up with, but just not yet. I would suggest merging this as is for now (after fixing the test back), and revisit it with the new dispatcher interface. Also note that the initialization **is** implicit, just deferred a bit.

**Comment by dom96** [comment]
Alright. Let's hope the behaviour of futures with no dispatcher doesn't catch people out. Otherwise we'll have to make that an error.

**Comment by yglukhov** [comment]
Thanks @zielmicha, @dom96! 👍
